dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java

```diff
@@ -274,177 +293,170 @@ private void handlePotentialFormatEntry(Set<FileFormat> fileFormats, Object v) {
-                        processDependencies(nodePrefix, endDependencies, sources);
-                    }
+                if (secondaryFile != null) {
+                    nodeDockerInfo.put(workflowStepId, new MutableTriple<>(secondaryFile, stepDockerRequirement, dockerUrl));
+                } else {
+                    nodeDockerInfo.put(workflowStepId, new MutableTriple<>(mainDescriptorPath, stepDockerRequirement, dockerUrl));
+                }
 
-                    toolInfoMap.put("UniqueEndKey", new ToolInfo(null, endDependencies));
-                    nodePairs.add(new MutablePair<>("UniqueEndKey", ""));
+            }
 
-                    // connect start node with them
-                    for (Pair<String, String> node : nodePairs) {
-                        if (toolInfoMap.get(node.getLeft()) == null) {
-                            toolInfoMap.put(node.getLeft(), new ToolInfo(null, Lists.newArrayList("UniqueBeginKey")));
-                        }
-                    }
-                    nodePairs.add(new MutablePair<>("UniqueBeginKey", ""));
+            if (type == LanguageHandlerInterface.Type.DAG) {
+                // Determine steps that point to end
+                List<String> endDependencies = new ArrayList<>();
 
-                    return Optional.of(setupJSONDAG(nodePairs, toolInfoMap, stepToType, nodeDockerInfo));
-                } else {
-                    return Optional.of(getJSONTableToolContent(nodeDockerInfo));
+                for (WorkflowOutputParameter workflowOutputParameter : workflow.getOutputs()) {
+                    Object sources = workflowOutputParameter.getOutputSource();
+                    processDependencies(nodePrefix, endDependencies, sources);
                 }
-            } catch (JsonParseException ex) {
-                LOG.error("The JSON file provided is invalid.", ex);
-                return Optional.empty();
+
+                toolInfoMap.put("UniqueEndKey", new ToolInfo(null, endDependencies));
+                nodePairs.add(new MutablePair<>("UniqueEndKey", ""));
+
+                // connect start node with them
+                for (Pair<String, String> node : nodePairs) {
+                    if (toolInfoMap.get(node.getLeft()) == null) {
+                        toolInfoMap.put(node.getLeft(), new ToolInfo(null, Lists.newArrayList("UniqueBeginKey")));
+                    }
+                }
+                nodePairs.add(new MutablePair<>("UniqueBeginKey", ""));
+
+                return Optional.of(setupJSONDAG(nodePairs, toolInfoMap, stepToType, nodeDockerInfo));
+            } else {
+                return Optional.of(getJSONTableToolContent(nodeDockerInfo));
             }
-        } else {
-            return Optional.empty();
+        } catch (ClassCastException | YAMLException | JsonParseException ex) {
+            throw new CustomWebApplicationException(this.CWL_PARSE_ERROR + ex.getMessage(), HttpStatus.SC_BAD_REQUEST);
```

Comment:
Log the exception

Reply:
[Logged](https://github.com/dockstore/dockstore/blob/c5277c33e2a3bf271f3be4e33b742fb646b1a076/dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java#L463).